keps/168-pending-workloads-visibility/README.md

```diff
@@ -0,0 +1,694 @@
+
+This risk exists because the operation requires iteration over the contents of
+the cluster queue, which requires a read lock on the queue. Also, positional
+changes to the list of pending workloads may require more frequent updates if
+attempt to keep the information up-to-date.
+
+In order to mitigate the risk we maintain the statuses on best-effort basis,
+and issue at most one update request in a configured interval,
+see [throttling of status updates](#throttling-of-status-updates).
+
+Additionally, we take periodically an in-memory snapshot of the ClusterQueue to
+allow generation of the status with `MaxCount` elements for LocalQueues and
+ClusterQueues without taking the read lock for a prolonged time:
+[In-memory snapshot of the ClusterQueue](#In-memory snapshot of the ClusterQueue).
+
+#### Large number of API requests triggered after workload admissions
+
+In a scenario when we have multiple LocalQueues pointing to the same
+ClusterQueue a workload that is admitted in one LocalQueue shifts positions of
+pending workloads in other LocalQueues. In the worst case scenario updating the
+LocalQueue statuses with new positions requires as many API requests as the
+number of LocalQueues. In particular, sending over 100 requests after workload
+admission would degrade Kueue performance.
+
+First, we propose to batch the LocalQueue updates by time intervals. This helps
+to avoid sending API requests per LocalQueue if the positions are shifted
+multiple times in a short period of time.
+
+Second, we introduce the `MaxPosition` parameter. With this parameter, the
+number of LocalQueues requiring an update can be controlled, because only
+LocalQueues with workloads at the top positions require an update.
+
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+The APIs of the status for LocalQueue and ClusterQueue are extended by
+structures which contain the list of pending workloads. In case of LocalQueue,
+also the workload position in the ClusterQueue is exposed.
+
+Updates to the structures are throttled, allowing for at most one update within
+a configured interval. Additionally, we take periodically an in-memory snapshot
+of the ClusterQueue.
+
+### Local Queue API
+
+```golang
+// LocalQueuePendingWorkload contains the information identifying a pending
+// workload in the local queue.
+type LocalQueuePendingWorkload struct {
+	// Name indicates the name of the pending workload.
+	Name string
+
+	// Position indicates the position of the workload in the cluster queue.
+	Position *int32
+
+	// MaxPosition indicates the maximal position of the workload in the cluster
+	// queue returned in the head.
+	MaxPosition *int32
```

Comment:
Did you mean to put this in Configuration?

Reply:
yes, fixed